core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt

```diff
@@ -0,0 +1,756 @@
+private fun getBlockedGauge(
+    rjsTrackSection: RJSTrackSection
+): DistanceRangeMap<LoadingGaugeConstraint> {
+    // This method has a bad complexity compared to more advanced solutions,
+    // but we don't expect more than a few ranges per track-section.
+    // TODO: use an interval tree
+    val res = distanceRangeMapOf<LoadingGaugeConstraint>()
+    if (
+        (rjsTrackSection.loadingGaugeLimits == null) ||
+            (rjsTrackSection.loadingGaugeLimits.isEmpty())
+    )
+        return res
+
+    // Sorts and removes duplicates
+    val transitions = TreeSet<Double>()
+    for (range in rjsTrackSection.loadingGaugeLimits) {
+        transitions.add(range.begin)
+        transitions.add(range.end)
+    }
+
+    val transitionsList = ArrayList(transitions) // Needed for index based loop
+    for (i in 1 until transitionsList.size) {
+        val begin = transitionsList[i - 1]
+        val end = transitionsList[i]
+        val allowedTypes = HashSet<RJSLoadingGaugeType>()
+        for (range in rjsTrackSection.loadingGaugeLimits) {
+            if (range.begin <= begin && range.end >= end) {
+                val compatibleTypes = compatibleGaugeTypeMap[range.category]
+                if (compatibleTypes == null) {
+                    logger.warn(
+                        "Invalid gauge type ${range.category} for track ${rjsTrackSection.id}"
+                    )
+                    allowedTypes.addAll(allLoadingGaugeTypeSet)
+                } else {
+                    allowedTypes.addAll(compatibleTypes)
+                }
+            }
+        }
+        res.put(begin.meters, end.meters, fromAllowedSet(allowedTypes))
+    }
+    return res
+}
+
+private fun buildCompatibleGaugeTypesMap(): Map<RJSLoadingGaugeType, Set<RJSLoadingGaugeType>> {
+    val compatibleGaugeType = hashMapOf<RJSLoadingGaugeType, Set<RJSLoadingGaugeType>>()
+    for (gaugeType in enumValues<RJSLoadingGaugeType>()) {
+        val compatibleTypes = gaugeType.compatibleGaugeTypes
+        if (compatibleTypes != null) compatibleGaugeType[gaugeType] = compatibleTypes
+    }
+    return compatibleGaugeType
+}
+
+val compatibleGaugeTypeMap = buildCompatibleGaugeTypesMap()
+val allLoadingGaugeTypeSet = enumValues<RJSLoadingGaugeType>().toSet()
+
+private fun parseRjsRouteWaypoint(
+    rjsDetector: RJSRouteWaypoint,
+    trackSectionNameToDistanceSortedDetectors:
+        MutableMap<String, TreeMap<Offset<TrackSection>, MutableList<String>>>,
+) {
+    val trackDetectors =
+        trackSectionNameToDistanceSortedDetectors.computeIfAbsent(rjsDetector.track) { TreeMap() }
+    val signalPosition = Offset<TrackSection>(rjsDetector.position.meters)
+    val detectorNames = trackDetectors.computeIfAbsent(signalPosition) { mutableListOf() }
+    detectorNames.add(rjsDetector.id)
+}
+
+/** Build detection zones on a partially-filled builder (track-sections, nodes and detectors) */
+private fun buildZones(builder: RawInfraFromRjsBuilder) {
```

Comment:
This whole function is quite long, and it's not immediately clear to me what every step does (or rather, why)

Reply:
I've added steps in an attempt to make it clearer